### DISCUSSION TITLE
Git operations slow compared to WSL2 on same machine

### DISCUSSION THREAD

**Discussion by HenryGessau** [git-for-windows/git#5227] [General]
If I clone a repo in the provided bash shell in Windows 11:
```
$ time __git_ps1
 (master|u=)
real    0m0.574s
user    0m0.000s
sys     0m0.030s
```

If I clone the same repo in WSL2 (Ubuntu) on the same machine:
```
$ time __git_ps1
 (master|u=)
real    0m0.032s
user    0m0.024s
sys     0m0.009s
```

Why is the git performance so much slower in the git-for-windows environment?

**Comment by dscho** [comment]
The problem is Git's over-reliance on multi-process operations. Basically, whenever you call `git.exe`, you pay a comparatively minor price on Linux because its `fork()` syscall is pretty quick, while the equivalent `CreateProcess()` takes much longer.

It gets _even_ worse when you do that in Git Bash, as the `__git_ps1` _Bash_ function does: In this environment, the processes are actually created via a `fork()` syscall _emulated by the POSIX emulation layer_ (also known as "the MSYS2 runtime", which is a friendly fork of the [Cygwin](https://cygwin.com) runtime). This not only pays the price of a `CreateProcess()` call, it actually pays it _twice_ because Unix-style signal handling is emulated for non-MSYS2 processes (such as `git.exe`) via a _second_ process, and to add insult to injury, the `fork()` syscall has to _copy_ the memory from the parent process and re-open file handles to imitate what `fork()` does on Unix. This is particularly painful because the only thing Bash does right after the POSIX emulation went through all of those steps _is to toss all of that away by calling `exec()` and replacing everything with the new `git.exe` process.

You can see all of those processes being spawned if you call `set -x` and after that simply hit <kbd>Enter</kbd>.

So yeah, it's quite slow, and it's a shame.

**Reply by HenryGessau** [reply]
Thanks for the detailed explanation.
